snapshot/git/gitdb/bundlestore.go

```diff
@@ -0,0 +1,200 @@
+package gitdb
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+
+	snap "github.com/scootdev/scoot/snapshot"
+	"github.com/scootdev/scoot/snapshot/bundlestore"
+)
+
+// BundlestoreConfig defines how to talk to Bundlestore
+type BundlestoreConfig struct {
+	Store bundlestore.Store
+}
+
+type bundlestoreBackend struct {
+	cfg *BundlestoreConfig
+}
+
+const bundlestoreIDText = "bs"
+
+// "bs-gc-<bundle>-<stream>-<sha>"
+
+func (b *bundlestoreBackend) parseID(id snap.ID, kind snapshotKind, extraParts []string) (snapshot, error) {
+	if len(extraParts) != 3 {
+		return nil, fmt.Errorf("cannot parse snapshot ID: expected 5 parts in bundlestore id: %s", id)
+	}
+	bundleName, streamName, sha := extraParts[0], extraParts[1], extraParts[2]
+
+	if err := validSha(sha); err != nil {
+		return nil, err
+	}
+
+	return &bundlestoreSnapshot{kind: kind, sha: sha, bundleName: bundleName, streamName: streamName}, nil
+}
+
+func (b *bundlestoreBackend) upload(s snapshot, db *DB) (snapshot, error) {
+	// We only have to upload a localSnapshot
+	switch s := s.(type) {
+	case *tagsSnapshot:
+		return s, nil
+	case *streamSnapshot:
```

Comment:
Still not seeing why tags can't be folded into the stream concept. Tags are a stream that doesn't advance and can't be updated with new data. But they both do the same core thing which is push/pull from repos.

I'm not sure you addressed stream uploading - what's pushing to master in your last sentence? And what's creating the ID? What is the user doing directly with git and also via our CLI?